profiling/src/lib.rs

```diff
@@ -0,0 +1,713 @@
+
+    // Set up runtime-id for Code Hotspots
+    RUNTIME_ID = profiling::Uuid::from(uuid::Uuid::new_v4());
+
+    ZendResult::Success
+}
+
+extern "C" fn ginit(global: *mut c_void) {
+    #[cfg(debug_assertions)]
+    {
+        logging::log_init(LevelFilter::Trace);
+        trace!("GINIT({:p})", global);
+    }
+
+    // Safety: The engine has allocated the globals to the right size and
+    // alignment, so it is safe to dereference this, but only when cast to
+    // MaybeUninit, as it's uninitialized (which is the point of this hook).
+    unsafe {
+        let global = global as *mut MaybeUninit<DatadogPhpProfilingGlobals>;
+        (*global).write(DatadogPhpProfilingGlobals::default());
+    }
+}
+
+unsafe extern "C" fn gshutdown(_global: *mut c_void) {
+    #[cfg(debug_assertions)]
+    trace!("GSHUTDOWN({:p})", _global);
+}
+
+extern "C" fn prshutdown() -> ZendResult {
+    #[cfg(debug_assertions)]
+    trace!("PRSHUTDOWN");
+
+    ZendResult::Success
+}
+
+/// The version of PHP at runtime, not the version compiled against. Sent as
+/// a profile tag.
+static mut PHP_VERSION: Option<String> = None;
+
+static mut PROFILER: RefCell<Option<Profiler>> = RefCell::new(None);
+
+/// The runtime ID, which is basically a universally unique "pid", so it can
+/// change on minit, as well as theoretically on fork. It should only be
+/// changed from thread-safe contexts.
+/// todo: support forking.
+static mut RUNTIME_ID: profiling::Uuid = profiling::Uuid::new();
+
+/// The version of the ZendEngine at runtime, not the version compiled against.
+/// It's currently unused, but I hope to send it as a metric or something soon
+/// so I'm keeping it here so I don't have to look up how to get it again.
+static mut ZEND_VERSION: Option<String> = None;
+
+/// # Safety
+/// Only call during an active request from the PHP thread!
+unsafe fn getenv(name: &CStr) -> Option<String> {
+    // CStr doesn't have a len() so turn it into a slice.
+    let name = name.to_bytes();
+
+    let val = sapi_getenv(name.as_ptr() as *const c_char, name.len() as c_ulong);
+    let val = val.into_string();
+    if !val.is_empty() {
+        return Some(val);
+    }
+
+    // If the sapi didn't have an env var, try the libc.
+    let val = libc::getenv(name.as_ptr() as *const c_char);
+    if val.is_null() {
+        return None;
+    }
+    let val = CStr::from_ptr(val);
+    return Some(String::from_utf8_lossy(val.to_bytes()).into_owned());
+}
+
+fn parse_boolean(mut string: String) -> Option<bool> {
+    // no boolean strings are longer than 5 characters
+    if string.is_empty() || string.len() > 5 {
+        return None;
+    }
+
+    string.make_ascii_lowercase();
+    match string.as_str() {
+        "1" | "on" | "yes" | "true" => Some(true),
+        "0" | "no" | "off" | "false" => Some(false),
+        _ => None,
+    }
+}
+
+/// Intern the string for the duration of a request.
+/// # Safety
+/// Do not use the string for more than the request lifetime!
```

Comment:
Would it be possible to enforce this with some lifetime on a (perhaps even empty) Request struct?

(It may very well not be, since request init and teardown are two separate FFI calls and rustc can't really know much about what's happening behind the scenes)